### ISSUE TITLE
Support multiple container formats

### ISSUE THREAD

**Issue by mattmoor** [bazelbuild/rules_docker#110]
I'm opening this issue to discuss options for supporting multiple container formats (OCI, RKT) via some of the core logic in `docker_build`.

@guymers @dlorenc @jonjohnsonjr FYI

**Comment by mattmoor** [comment]
This is something I've been pondering for some time, and so I have some thoughts, which I'll write down to kick things off. :)

At least for OCI vs. Docker v2.2, the differences are currently cosmetic (trivial schema changes and different mime types).  Moreover, the [new intermediate format](https://github.com/bazelbuild/rules_docker/blob/master/intermediate-form.md) for `rules_docker` doesn't know anything about mimes.

So I'm wondering if we want the evolution of things to be...

### container_image

Add support for a generic `container_image` rule, which builds our intermediate format and is agnostic of the final format.
```python
container_image(
    name = "foo",
    base = "@bar//image",
    ...
)
```
We can achieve this by either computing all forms simultaneously (and letting Bazel prune extraneous actions), or using a container agnostic form and projecting it into a particular format on consumption.  We could support form-specific functionality, but consuming that through `container_image` should warn.

These rules couldn't preserve the current `incremental_load.sh.tpl` or `:foo.tar` functionalities without a hint as to which what form the user wants things to default.  I think we could expose a kwarg for the default format, which makes the next part trivial.

I think moving this core definition into `bazelbuild/rules_container` makes complete sense, and we can hollow out `bazelbuild/rules_docker` to proxy to our core implementation.

### {docker,oci,rkt}_image

For clarity / brevity, I think it makes to wrap `container_image` with a default format with a name to reflect that specific form.  This should suppress form-specific warnings (for our form) or turn them into errors (other forms).

### {docker,oci,rkt}_pull

As today, these should pull a particular form into our agnostic intermediate format.  If needed, we can leverage the bridging utilities in `google/containerregistry`.

### {docker,oci,rkt}_push

As today, these should exit our agnostic intermediate format, and publish the image in the particular format.

### Open Questions

* Can we express everything we want via an agnostic intermediate format?
* How would `rules_k8s` (still a prototype) build on top of an agnostic format?
* Do we want/need a `container_bundle` (a la `docker_bundle`)?

**Comment by guymers** [comment]
I like the idea of a generic `container_image` rule that could build different formats. Would it be worth it to maintain specific image rules for each form? Are there any form specific values outside of the OCI image spec that are useful enough to support? 

`rules_k8s` looks like a very nice tool. From a quick glance it looks like it basically just needs to be able to push images? So if it just delegated to some sort of `container_push` rule it shouldn't need to care what the format is? That push rule could use authentication information to know what format to use?

I am not 100% sure on when you need to use `docker_bundle`? Is it an optimisation that effectively de-duplicates shared layers to allow push to be more efficient?

**Comment by mattmoor** [comment]
In my perfect world, we have `docker_build`, but it's a thin wrapper.  Without a compelling reason, I'm not sure we want them to have any substance.  Certainly v2.2 and OCI are too similar to have meaningful feature divergence.  I could see rkt having some potentially, but I'm not familiar enough with rkt to say.

`rules_k8s` does need to be able to push, but I'm not sure what you mean by relying on the authentication information to determine the format?  At a minimum we could attempt to push each format, and fallback on alternatives.

`docker_bundle` is a convenience for synthesizing a set of images into a single `docker save`-style tarball with user specified aliases (unlike `docker_build`, which just provides temporary names).  It isn't an optimization, just a convenience for moving images around without using a registry (there are some scenarios where the K8s team does this).

**Comment by mattmoor** [comment]
re: divergence

That's not to say the forks won't diverge down the road, I'm just not aware of any today.

I think we could add form-specific stuff into `foo_build`, and enable down-conversion (`container => foo`), but not up-conversion (`foo => container`).  As a (reformed) compiler guy, I think of dataflow lattices.
